maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java

```diff
@@ -242,21 +274,17 @@ public DefaultGraphBuilder( BuildResumptionDataRepository buildResumptionDataRep
     {
         List<MavenProject> result = projects;
 
-        if ( !request.getExcludedProjects().isEmpty() )
+        ProjectActivation projectActivation = request.getProjectActivation();
+        Set<String> requiredSelectors = projectActivation.getRequiredInactiveProjectSelectors();
+        Set<String> optionalSelectors = projectActivation.getOptionalInactiveProjectSelectors();
+        if ( !requiredSelectors.isEmpty() || !optionalSelectors.isEmpty() )
         {
-            File reactorDirectory = getReactorDirectory( request );
+            Set<MavenProject> excludedProjects = new HashSet<>( requiredSelectors.size() + optionalSelectors.size() );
+            excludedProjects.addAll( getProjectsBySelectors( request, projects, requiredSelectors, true ) );
+            excludedProjects.addAll( getProjectsBySelectors( request, projects, optionalSelectors, false ) );
```

Comment:
Call me blind, but this just looks like [here](https://github.com/apache/maven/pull/446/files#diff-5a5fe4cfc363538b43d7a3e89cc8b12b64d7ae0f4159caa45740405a299b6e3eR180-R187), but method names are opposite. Can you explain please?

Reply:
This piece of code removes the excluded projects (`-pl !proj` and `-pl -?proj`) from the reactor. The other method limits the reactor to all selected projects (`-pl proj,?proj`). Both methods find all projects by the project selectors and by whether they are required or optional.

Reply:
Interesting, thanks. What will happen with `-pl !proj,+proj`?

Reply:
Interesting question, the `DefaultGraphBuilder` will first filter based on the selected modules (+), afterwards it will filter out the excluded projects, so in this case nothing will be selected. Actually, it will result in a bug (I'll create a JIRA ticket for it)

This is an attempt in the integration-test project:
```
>mvn validate -pl !core-it-suite,+core-it-suite
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.630 s
[INFO] Finished at: 2021-02-14T20:15:52+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] No goals have been specified for this build. You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-reso
urces, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean,
 clean, post-clean, pre-site, site, post-site, site-deploy. -> [Help 1]```

Reply:
Actually.. this is the case for 3.6.3, the latest master actually builds the targeted project just fine. I'm not sure it's not filtered out... Let me get back to that.

Reply:
OK...

Reply:
I need a bit more time for this. The latest maven master build actually has the same bug as 3.6.3, but the build of this PR does not have the error. I need to go now, but will check tomorrow or something if I can find out why (and whether we want it or not..?)

Reply:
Right, the data structure inside `ProjectActivation` is a `Map<(ProjectSelector)String, (enum)ActivationSetting>`, so that means that a project cannot be both turned selected and excluded. The one who gets set later wins (in `-pl !a,+a` '+a' wins).
This was not fully intentional, but I don't dislike the outcome. The same goes for ProfileActivation, by the way.

The bug remains though if the user invokes `mvn -pl -a,+:a` (same proj, different selector). It will output "No goals have been specified for this build.". I'll improve the error message in this case.

Reply:
But shouldn't that be a null sum game? "No goal ..." sounds wrong in any case, doesn't it?

Reply:
"No goal" is definitely a bug. I'm thinking of throwing an exception when the reactor command line flags result in an empty project list.

Reply:
but that would actually only take care of the symptom, wouldn't it?

Reply:
I guess it would, but I'm not sure how to fix it otherwise. What would you rather have? I don't think we can already check whether `a` or `:a` are the same in an earlier stage. 

I've just committed the exception before I read your reply. Please let me know what you think.

Reply:
That's acceptable. I think we need to place a comment above that checking saying it is a workaround and create an issue with the actual cause for later analysis because as soon as this goes to master it will be forgotten!

Reply:
I agree, but I don't think I understand how it would ideally work. Would you want to let `-a,+:a` work the same way as `-a,+a` instead of have the exception thrown? (last one wins)

Reply:
I can't tell what the correct behavior would be, but not saying no goals given. I guess this is a separate discussion.

Reply:
True, but that's now fixed with this last commit I did, right? You will now get the following:
```
> mvn -pl !core-it-suite,:core-it-suite validate
[INFO] Scanning for projects...
[ERROR] The project exclusion settings resulted in an empty project list, please correct the usage of --projects/-pl !<project> or -<project>
```

Reply:
Fixed the symptom. That's what I understand and it should carry a comment about that. One should explore the "No goal..." issue later.

Reply:
We discussed this with @rfscholte and the idea is to:

1. If the user gives conflicting settings for one project, they will cancel each other out: `-pl :a,-:a,:b` will yield selection of `:b`.
2. When the above rule yields no selected projects (everything is cancelled out), we throw ["empty project list" message](https://github.com/apache/maven/pull/446#discussion_r577123679).
3. When `?:a,:a` is specified, `:a` wins because it's "stronger".
4. When `a,:a` is specified (two selectors that yield the same module) the above rules apply as they refer to the same module. It is not considered a failure or error.

What are your thoughts on this?

Reply:
All four steps make sense here. Though, I am still confused about the "No goals here...".

Reply:
The "No goals here" is in 3.6.3, IIUC. With these proposed changes it shouldn't happen anymore.

Reply:
But only because you throw the exception, right?

Reply:
Yes: because these changes detect when `-pl` leads to no modules being selected, and the `DefaultGraphBuilder` then aborts the process, we will no longer get the "No goals have been specified" message.

Reply:
@michael-o we have implemented all logic to make the 4 points in [the reply above](https://github.com/apache/maven/pull/446#discussion_r583679349) work. It's up for review again if you have some time :)